ecosystem/typescript/sdk_v2/src/bcs/deserializer.ts

```diff
@@ -0,0 +1,192 @@
+interface Deserializable<T> {
+  deserialize(deserializer: Deserializer): T;
+}
+
+export class Deserializer {
+  private buffer: ArrayBuffer;
+
+  private offset: number;
+
+  constructor(data: Uint8Array) {
+    // copies data to prevent outside mutation of buffer.
+    this.buffer = new ArrayBuffer(data.length);
+    new Uint8Array(this.buffer).set(data, 0);
+    this.offset = 0;
+  }
+
+  private read(length: number): ArrayBuffer {
+    if (this.offset + length > this.buffer.byteLength) {
+      throw new Error("Reached to the end of buffer");
+    }
+
+    const bytes = this.buffer.slice(this.offset, this.offset + length);
+    this.offset += length;
+    return bytes;
+  }
+
+  /**
+   * Deserializes a string. UTF8 string is supported. Reads the string's bytes length "l" first,
+   * and then reads "l" bytes of content. Decodes the byte array into a string.
+   *
+   * BCS layout for "string": string_length | string_content
+   * where string_length is a u32 integer encoded as a uleb128 integer, equal to the number of bytes in string_content.
+   *
+   * @example
+   * ```ts
+   * const deserializer = new Deserializer(new Uint8Array([8, 49, 50, 51, 52, 97, 98, 99, 100]));
+   * assert(deserializer.deserializeStr() === "1234abcd");
+   * ```
+   */
+  deserializeStr(): string {
+    const value = this.deserializeBytes();
+    const textDecoder = new TextDecoder();
+    return textDecoder.decode(value);
+  }
+
+  /**
+   * Deserializes an array of bytes.
+   *
+   * BCS layout for "bytes": bytes_length | bytes
+   * where bytes_length is a u32 integer encoded as a uleb128 integer, equal to the length of the bytes array.
+   */
+  deserializeBytes(): Uint8Array {
+    const len = this.deserializeUleb128AsU32();
+    return new Uint8Array(this.read(len));
+  }
+
+  /**
+   * Deserializes an array of bytes. The number of bytes to read is already known.
+   *
+   */
+  deserializeFixedBytes(len: number): Uint8Array {
+    return new Uint8Array(this.read(len));
+  }
+
+  /**
+   * Deserializes a boolean value.
+   *
+   * BCS layout for "boolean": One byte. "0x01" for true and "0x00" for false.
+   */
+  deserializeBool(): boolean {
+    const bool = new Uint8Array(this.read(1))[0];
+    if (bool !== 1 && bool !== 0) {
+      throw new Error("Invalid boolean value");
+    }
+    return bool === 1;
+  }
+
+  /**
+   * Deserializes a uint8 number.
+   *
+   * BCS layout for "uint8": One byte. Binary format in little-endian representation.
+   */
+  deserializeU8(): Uint8 {
+    return new DataView(this.read(1)).getUint8(0);
+  }
+
+  /**
+   * Deserializes a uint16 number.
+   *
+   * BCS layout for "uint16": Two bytes. Binary format in little-endian representation.
+   * @example
+   * ```ts
+   * const deserializer = new Deserializer(new Uint8Array([0x34, 0x12]));
+   * assert(deserializer.deserializeU16() === 4660);
+   * ```
+   */
+  deserializeU16(): Uint16 {
+    return new DataView(this.read(2)).getUint16(0, true);
```

Comment:
Is DataView platform dependent?  How is little endian vs big endian?

Reply:
so normally `DataView` is big-endian, actually, but the second argument in `getUint16` (in our case, `true`) is the `littleEndian` flag.